File: scripts/expenses.R
# expenses.R
# prep the expenses data and push to Socrata
# Adam Scherling
# Written April 18, 2018. Updated June 11, 2018.

## Setup

# load required packages
if (!require("pacman")) install.packages("pacman")
pacman::p_load(dplyr, magrittr, tidyr, lubridate, data.table, RSocrata, readxl)

# set directories
# data = directory where the data are saved
# login = directory where Socrata login information is saved
#         (username.txt, password.txt - these should be plain text files 
#		   containing nothing but the username and password, respectively)
dir <- list(data="~/github/laBudget/data/approved_budget/FY18-19/",
       login="~/github/laBudget/scripts/")

# set the working directory
setwd(dir$data)

# url for the dataset
socrata_url <- "https://data.lacity.org/A-Prosperous-City/Open-Budget-Appropriations-Fiscal-Years-2010-2019/5242-pnmt"

# API endpoint for the dataset
# socrata_endpoint <- "https://data.lacity.org/resource/ws4f-n5ax.json"
socrata_endpoint <- "https://data.lacity.org/resource/5242-pnmt.json"

# filename of Excel spreadsheet
excel_file <- "Expenditures_Sec2 All Regular Depts and NonDepts_1819Adopted.xlsx"

# get Socrata username and password for upload
username <- readLines(paste0(dir$login, 'username.txt'))
password <- readLines(paste0(dir$login, 'password.txt'))

# get a timestamp
timestamp <- gsub(':','.',now())
timestamp <- gsub(' ','_',timestamp)

# prevent R from writing to scientific notation
options(scipen=999)

# get the fiscal year being updated
# default to the year after the current date
new_fiscal_year <- year(today()) + 1



## Back up existing data

# read in the existing data on Socrata
old_expenses <- read.socrata(socrata_url)

# save a copy as a backup
write.csv(old_expenses, paste0('old_expenses_',timestamp,'.csv'), row.names=F)

# filter out any data from the fiscal year that's being updated
old_expenses %<>% filter(Fiscal_Year!=new_fiscal_year)



## Read in new data

# read in the new expenses data. section 2 only
new_expenses <- read_excel(excel_file) %>% data.frame

# keep only the desired columns
# assumes that the last column contains the data of interest (correct fiscal year, proposed/adopted as desired)
new_expenses %<>% select(Dept.Code, Dept.Name, Org.Level.5.Code, Org.Level.5.Name, Prog.Code, Prog.Name,
	Source.Fund.Code, Source.Fund.Name, Account.Code, Account.Name, 
	colnames(new_expenses)[ncol(new_expenses)])

# change the column names of the new data to match the old data
# missing columns: Program_Priority, Expense_Type
colnames(new_expenses) <- c("Dept_Code","Department_Name","SubDept_Code",
	"SubDepartment_Name","Prog_Code","Program_Name","Source_Fund_Code",
	"Source_Fund_Name","Account_Code","Account_Name","Appropriation")

# remove rows with no appropriation data
new_expenses %<>% filter(!is.na(Appropriation))




## Get Program Priorities
# assign a category to each expenditure, e.g. A Well Run City, A Safe City,...
# an attempt is made to do this by matching the categories in the old dataset
# to new observations with the same department and program codes.
# unfortunately this doesn't work all that well

# for existing program names, get the program priority
priority <- old_expenses %>% select(Dept_Code, Prog_Code, Program_Priority)
priority <- priority[!duplicated(priority),]
priority <- na.omit(priority)

# remove duplicated Dept_Code, Prog_Code combinations. 
# somewhat undiscriminating, can be refined later
dd <- priority %>% select(Dept_Code, Prog_Code) %>% duplicated
priority <- priority[!dd,]

# merge in the program priority and expense type
new_expenses <- merge(new_expenses, priority, by=c("Dept_Code","Prog_Code"), all.x=T)

# examine the merge
# sapply(colnames(new_expenses), function(j) sum(is.na(new_expenses[,j])))



# alternatively: assign program priority using the descriptions dataset
priority2 <- read.csv('program_priorities.csv', stringsAsFactors=F) %>% select(Prog_Code, Program_Priority)
new_expenses <- merge(new_expenses, priority2, by="Prog_Code", all.x=T)

# merge the two estimates of priority, giving preference to the one from the descriptions dataset
new_expenses$Program_Priority <- ifelse(is.na(new_expenses$Program_Priority.y) | 
	new_expenses$Program_Priority.y=="", new_expenses$Program_Priority.x, new_expenses$Program_Priority.y)

new_expenses$Program_Priority[is.na(new_expenses$Program_Priority)] <- "Not Categorized"

# what portion of the appropriations have been assigned a program priority?
tapply(new_expenses$Appropriation, new_expenses$Program_Priority, sum) / sum(new_expenses$Appropriation)



### Miscellaneous data transformations


# add a fiscal year column.
new_expenses$Fiscal_Year <- new_fiscal_year

# add a blank expense_type column. this is no longer provided
new_expenses$Expense_Type <- NA

# arrange columns in the same order as the old expenses data
new_expenses %<>% select(colnames(old_expenses))

# convert appropriation column to character
new_expenses %<>% mutate(Appropriation=as.character(Appropriation))

# combine the old and new data
expenses <- rbind(new_expenses, old_expenses)

# sort according to fiscal year, department name, program name, account name
expenses %<>% arrange(desc(Fiscal_Year), Department_Name, Program_Name, Account_Name)


# write to csv
write.csv(expenses, 'expenses.csv', row.names=F)

# upload the data to Socrata
write.socrata(dataframe = expenses,
              dataset_json_endpoint = socrata_endpoint,
              update_mode = "REPLACE",
              email = username,
              password = password)

File: scripts/python-scripts/expenses.py
#!/usr/bin/env python3
# expenses.py
# prep the expenses data and push to Socrata
# Adam Scherling April 18, 2018. Updated June 11, 2018
# Converted/Updated July 2021, Irene Tang


####################
## Setup
####################

# make sure to install these packages before running:
# pip install pandas
# pip install sodapy

import datetime
import pandas as pd
import numpy as np
import credentials
from sodapy import Socrata

# url for the dataset
socrata_url = 'https://data.lacity.org/A-Prosperous-City/Open-Budget-Appropriations-Fiscal-Years-2010-2019/5242-pnmt'

# API endpoint for the dataset
socrata_endpoint = 'https://data.lacity.org/resource/5242-pnmt.json'

# Identifier for the dataset
socrata_identifier = '5242-pnmt'

# set up Socrata client
client = Socrata('data.lacity.org', None)

# uncomment if you are going to log in / push to the data portal
# username = credentials.lahub_user
# password = credentials.lahub_pass
# apptoken = credentials.lahub_auth
# client = Socrata('data.lacity.org', apptoken, username=username, password=password)

# get the fiscal year being updated; default to the year after the current date
timestamp = datetime.datetime.now()
current_year = timestamp.year
new_fiscal_year = current_year + 1

####################
## Read in existing data
####################
filepath_prefix = '../../data/approved_budget/FY21-22/'

# read in the existing data on Socrata
old_expenses = pd.DataFrame.from_records(client.get(socrata_identifier, limit=99999999999999))

# save a copy as a local backup -- especially before pushing the output back to overwrite the existing data on Socrata
old_expenses.to_csv(
    f'{filepath_prefix}old_expenses_{timestamp}.csv', index=False)

# filter out any data from the fiscal year that's being updated (only keep the rows where fiscal_year!=new_fiscal_year)
old_expenses = old_expenses[old_expenses['fiscal_year']!=new_fiscal_year]

####################
## Read in new data
####################

# read in the new expenses data. section 2 only
csv_file = '../../data/approved_budget/FY21-22/Expenditures_Sec2_for_2122_Adopted.csv'
new_expenses = pd.read_csv(csv_file)

# keep only the desired columns
# assumes that the last column contains the data of interest (correct fiscal year, proposed/adopted as desired)
appropriation = new_expenses.columns[-1]  # e.g. '2021-22 Adopted'
new_expenses = new_expenses[['Dept Code', 'Dept Name', 'Org Level 5 Code', 'Org Level 5 Name', 'Prog Code', 'Prog Name', 'Source Fund Code', 'Source Fund Name', 'Account Code', 'Account Name', appropriation]]

# rename the columns of the new data to match the old data
# missing columns: Program_Priority, Expense_Type
new_expenses.columns = ['dept_code', 'department_name', 'subdept_code', 'subdepartment_name', 'prog_code', 'program_name', 'source_fund_code', 'source_fund_name', 'account_code', 'account_name', 'appropriation']

# remove rows with no appropriation data
new_expenses.dropna(how='all', subset=['appropriation'], inplace=True)



####################
## Get Program Priorities
# assign a category to each expenditure, e.g. A Well Run City, A Safe City
####################

## first, an attempt is made to do this by matching the categories in the old dataset to new observations with the same department and program codes. unfortunately this doesn't work all that well
priorities1 = old_expenses[['dept_code', 'prog_code', 'program_priority']]
# TODO unsure about the logic of this line? it was in the original .R so I kept it here commented-out
# priorities1 = priorities1.drop_duplicates(subset='program_priority', keep='first')
priorities1.dropna(inplace=True)

# remove duplicate Dept_Code, Prog_Code combinations
priorities1.drop_duplicates(subset=['dept_code', 'prog_code'], keep='first', inplace=True)

## merge in the program_priorities1
new_expenses = pd.merge(new_expenses, priorities1, on=['dept_code', 'prog_code'], how='left')
new_expenses.loc[new_expenses['program_priority']=='','program_priority']='Not Categorized'
new_expenses['program_priority'].fillna('Not Categorized', inplace = True)

new_expenses = new_expenses.astype({'prog_code': str})

## second, assign program priority using the descriptions dataset
program_priorities2_filename = '../../data/approved_budget/FY18-19/program_priorities.csv'
priorities2 = pd.read_csv(program_priorities2_filename)

priorities2.dropna(inplace=True)
priorities2 = priorities2[['Prog_Code', 'Program_Priority']]

priorities2.columns = [x.lower() for x in priorities2.columns]
priorities2 = priorities2.astype({'prog_code': str})


## merge in the program_priorities2 -- this will create two columns: program_priority_x, program_priority_y (one for each attempt)
new_expenses = pd.merge(new_expenses, priorities2, on='prog_code', how='left')

## consolidate the two estimates of priority, giving preference to the one from the descriptions dataset
new_expenses['program_priority'] = np.where(new_expenses.program_priority_y, new_expenses.program_priority_y, new_expenses.program_priority_y)

# fill the empty slots with 'Not Categorized'
new_expenses['program_priority'].fillna('Not Categorized', inplace = True)
new_expenses.loc[new_expenses['program_priority']=='','program_priority']='Not Categorized'

# delete columns program_priority_x, program_priority_y
new_expenses.drop(['program_priority_x', 'program_priority_y'], axis=1, inplace=True)



####################
### Miscellaneous data transformations
####################

# add a fiscal year column
new_expenses['fiscal_year'] = new_fiscal_year

# add a blank expense_type column. this is no longer provided
new_expenses['expense_type'] = None

# arrange columns in the same order as the old expenses data
new_expenses = new_expenses[old_expenses.columns]

# convert appropriation column to character
# new_expenses.assign(Appropriation=lambda x: str(x))

# combine the old and new data
expenses = pd.concat([new_expenses, old_expenses], axis=0)

# sort according to fiscal year, department name, program name, account name
expenses = expenses.astype({'fiscal_year': int})
expenses.sort_values(by=['fiscal_year'], ascending=False)
expenses.sort_values(by=['department_name', 'program_name', 'account_name'], ascending=True)

####################
## Save data
####################

# write to csv
expenses.to_csv(f'{filepath_prefix}new_expenses.csv')

# upload the data to Socrata
# client.replace(socrata_identifier, expenses)
